guide on the usage of docker (#1111)
microsoft/autogen: website/docs/Installation.md

```
docker build -f samples/dockers/Dockerfile.base -t autogen_img https://github.com/microsoft/autogen.git#main
```
which includes some common python libraries and essential dependencies of AutoGen, or build from `Dockerfile.full` which include additional dependencies for more advanced features of AutoGen with the following command line:

```
docker build -f samples/dockers/Dockerfile.full -t autogen_full_img https://github.com/microsoft/autogen.git
```
Once you build the docker image, you can use `docker images` to check whether it has been created successfully.

#### Step 3.  Run applications built with AutoGen from a docker image.

**Mount your code to the docker image and run your application from there:** Now suppose you have your application built with AutoGen in a main script named `twoagent.py` ([example](https://github.com/microsoft/autogen/blob/main/test/twoagent.py)) in a folder named `myapp`. With the command line below, you can mont your folder and run the application in docker.

```python
# Mount the local folder `myapp` into docker image and run the script named "twoagent.py" in the docker.

> docker run -it -v `pwd`/myapp:/myapp autogen_img:latest python /myapp/main_twoagent.py

Reviewer: I might be wrong because I didn't test it, but I believe that `git` will not work inside the container the way it has been set up. If you use ssh for authentication, then commands such as `git pull` and `git push` will look into `~/.ssh/` for auth data. We could provide instructions for mapping `.ssh` folder in host to container to avoid this.

# Installation

## Option 1: Install and Run AutoGen in Docker
#### Step 1. Install [docker](https://www.docker.com/).

Docker is recommended as a properly built docker image could provide isolated and consistent environment to run your code securely across platforms. Install docker following [this instruction](https://docs.docker.com/get-docker/).

#### Step 2. Build a docker image

> AutoGen provides [dockerfiles](https://github.com/microsoft/autogen/tree/main/samples/dockers/) that could be used to built docker images. Use the following command line to build a docker image named `autogen_img` (or some other name you like) one of the provided dockerfiles named `Dockerfile.baes`:

Reviewer 2: ```suggestion
AutoGen provides [dockerfiles](https://github.com/microsoft/autogen/tree/main/samples/dockers/) that could be used to built docker images. Use the following command line to build a docker image named `autogen_img` (or some other name you like) one of the provided dockerfiles named `Dockerfile.base`:
```

```

> docker build -f sample/docker/Dockerfile.base -t autogen_img https://github.com/microsoft/autogen.git

Reviewer 3: Note that this command is running when inside the root directory.

Author: Not necessarily? It will pull the dockerfile from the autgen github following the specified file path.

Reviewer 3: Oh right! I missed the github link.

```
docker build -f sample/docker/Dockerfile.full -t autogen_full_img https://github.com/microsoft/autogen.git
```
Once you build the docker image, you can use `docker images` to check whether it has been created successfully.

#### Step 3.  Run applications built with AutoGen from a docker image.

**Save keys into environment variable:**
If your application needs an oai key, save your oai key into an environment variable `OPENAI_API_KEY`, which can be done by adding the following line in your `~/.bashrc` (or `~/.zshrc`),

```
export OPENAI_API_KEY="sk-xxxxxx"
```

**Mount your code to the docker image and run your application from there:** Now suppose you have your application built with AutoGen in a folder named `myapp`. You can mount it into the docker image and run it. In the example below, the folder `myapp` is mounted into `/myapp` in the docker, and the script `main_twoagent.py` is executed in the docker.

```python
# Mount the local folder `myapp` into docker image and run the script in the docker.
docker run -it -e OPENAI_API_KEY=$OPENAI_API_KEY -v `pwd`/myapp:/myapp autogen_img:latest python /myapp/main_twoagent.py
```

> Similarly, you may also run [AutoGen Studio](https://github.com/microsoft/autogen/tree/main/samples/apps/autogen-studio) (assuming that you have built a docker image named `autogen_full_img` with `Dockerfile.full`)as below:

Reviewer 2: ```suggestion
Similarly, you may also run [AutoGen Studio](https://github.com/microsoft/autogen/tree/main/samples/apps/autogen-studio) (assuming that you have built a docker image named `autogen_full_img` with `Dockerfile.full`) as below:
```

```
docker run -it -e OPENAI_API_KEY=$OPENAI_API_KEY -p 8081:8081 autogen_full_img:latest autogenra ui --host 0.0.0.0
```

> Then open `http://localhost:8081/` in your browser to use AutoGen assistant.

Reviewer 2: ```suggestion
Then open `http://localhost:8081/` in your browser to use AutoGen Studio.
```

## Option 2: Install AutoGen Locally Using Virtual Environment

> When not using a docker container, we recommend using a virtual environment to install AutoGen. This will ensure that the dependencies for AutoGen are isolated from the rest of your system.

Reviewer 2: Double check this line. "while not using a docker container" can be confusing. Check line 135-142.

Author: Revised. Thanks!

# Installation

## Option 1: Install and Run AutoGen in Docker
#### Step 1. Install [docker](https://www.docker.com/).

Docker is recommended as a properly built docker image could provide isolated and consistent environment to run your code securely across platforms. Install docker following [this instruction](https://docs.docker.com/get-docker/).

For Mac users, alternatively you may choose to install [colima](https://smallsharpsoftwaretools.com/tutorials/use-colima-to-run-docker-containers-on-macos/) to run docker containers, if there is any issues with starting the docker daemon.

#### Step 2. Build a docker image

> AutoGen provides [dockerfiles](https://github.com/microsoft/autogen/tree/main/samples/dockers/) that could be used to built docker images. Use the following command line to build a docker image named `autogen_img` (or some other name you like) one of the provided dockerfiles named `Dockerfile.base`:

Reviewer 4: nit: 
```suggestion
AutoGen provides [dockerfiles](https://github.com/microsoft/autogen/tree/main/samples/dockers/) that could be used to build docker images. Use the following command line to build a docker image named `autogen_img` (or some other name you like) one of the provided dockerfiles named `Dockerfile.base`:
```

```
docker build -f samples/dockers/Dockerfile.base -t autogen_img https://github.com/microsoft/autogen.git
```
which includes some common python libraries and essential dependencies of AutoGen, or build from `Dockerfile.full` which include additional dependencies for more advanced features of AutoGen with the following command line:

```

> docker build -f sample/docker/Dockerfile.full -t autogen_full_img https://github.com/microsoft/autogen.git

Reviewer 5: 1. should be `samples/dockers/Dockerfile.full`.
2. I cannot run this command but the following works in autogen root directory.
`docker build -f sample/docker/Dockerfile.full -t autogen_full_img .`

Author: 1 addressed. Thanks!
2 since the folders are added in this PR and not yet in the main repo, the link is not working right now. It will work once this PR is merged.  This should work now: `docker build -f samples/dockers/Dockerfile.base -t autogen_img https://github.com/microsoft/autogen.git#add-docker`

```
docker build -f samples/dockers/Dockerfile.base -t autogen_img https://github.com/microsoft/autogen.git#main
```
which includes some common python libraries and essential dependencies of AutoGen, or build from `Dockerfile.full` which include additional dependencies for more advanced features of AutoGen with the following command line:

```
docker build -f samples/dockers/Dockerfile.full -t autogen_full_img https://github.com/microsoft/autogen.git
```
Once you build the docker image, you can use `docker images` to check whether it has been created successfully.

#### Step 3.  Run applications built with AutoGen from a docker image.

**Save keys into environment variable:**
If your application needs an oai key, save your oai key into an environment variable `OPENAI_API_KEY`, which can be done by adding the following line in your `~/.bashrc` (or `~/.zshrc`),

```
export OPENAI_API_KEY="sk-xxxxxx"

> ```

Reviewer 2: Shouldn't we recommend the approach of using the oai_config_list file?

**Mount your code to the docker image and run your application from there:** Now suppose you have your application built with AutoGen in a folder named `myapp`. You can mount it into the docker image and run it. In the example below, the folder `myapp` is mounted into `/myapp` in the docker, and the script `main_twoagent.py` is executed in the docker.

```python
# Mount the local folder `myapp` into docker image and run the script in the docker.
docker run -it -e OPENAI_API_KEY=$OPENAI_API_KEY -v `pwd`/myapp:/myapp autogen_img:latest python /myapp/main_twoagent.py
```

Similarly, you may also run [AutoGen Studio](https://github.com/microsoft/autogen/tree/main/samples/apps/autogen-studio) (assuming that you have built a docker image named `autogen_full_img` with `Dockerfile.full`) as below:

```

> docker run -it -e OPENAI_API_KEY=$OPENAI_API_KEY -p 8081:8081 autogen_full_img:latest autogenra ui --host 0.0.0.0

Reviewer 2: avoid exporting env var.
@victordibia is it avoidable for autogen studio?

Reviewer: Some tests fail if OPENAI_API_KEY is not defined. We should probably make those tests use OAI_CONFIG_LIST and then there is no need to have OPENAI_API_KEY defined.

deactivate
```

### Option b: conda

> Another option is with `Conda`, Conda works better at solving dependency conflicts than pip. You can install it by following [this doc](https://docs.conda.io/projects/conda/en/stable/user-guide/install/index.html),

Reviewer 2: ```suggestion
Another option is with `Conda`. You can install it by following [this doc](https://docs.conda.io/projects/conda/en/stable/user-guide/install/index.html),
```
We are not building a conda package, so the dependency conflict point is off.

### Optional Dependencies
- #### docker

> For the best user experience and seamless code execution, even if you are installing AutoGen locally, we highly recommend using Docker for [code execution](FAQ.md#enable-python-3-docker-image).

Reviewer 2: ```suggestion
Even if you install AutoGen locally, we highly recommend using Docker for [code execution](FAQ.md#enable-python-3-docker-image).
```
